File: api/src/services/neighbor-chunk.service.ts
import { DatabaseModel, SearchResult } from '../models/database.model';

// Back-compat for older code paths (e.g. chat.service.ts).
export function addNeighborChunks(topChunks: any[], allChunks: any[], window = 1) {
  const chunkMap = new Map<string, any>();

  allChunks.forEach(chunk => {
    const key = `${chunk.docId}_${chunk.order}`;
    chunkMap.set(key, chunk);
  });

  const result = new Map<number, any>();

  topChunks.forEach(chunk => {
    for (let offset = -window; offset <= window; offset++) {
      const key = `${chunk.docId}_${chunk.order + offset}`;
      const neighbor = chunkMap.get(key);
      if (neighbor) {
        result.set(neighbor.chunkId, neighbor);
      }
    }
  });

  return Array.from(result.values());
}

export class NeighborChunkService {
  constructor(private db: DatabaseModel) {}

  async expand(chunks: SearchResult[], window = 1): Promise<SearchResult[]> {
    const w = Number.isFinite(window) ? Math.max(0, Math.floor(window)) : 1;
    if (w === 0 || chunks.length === 0) return chunks;

    // Build + merge ranges per document to minimize DB queries.
    const byDoc = new Map<string, Array<{ start: number; end: number }>>();
    for (const c of chunks) {
      const start = Math.max(0, c.chunk_index - w);
      const end = c.chunk_index + w;
      const list = byDoc.get(c.document_id) || [];
      list.push({ start, end });
      byDoc.set(c.document_id, list);
    }

    const mergedRanges: Array<{ document_id: string; start_index: number; end_index: number }> = [];
    for (const [documentId, ranges] of byDoc.entries()) {
      ranges.sort((a, b) => a.start - b.start);
      let current = ranges[0];
      for (let i = 1; i < ranges.length; i++) {
        const r = ranges[i];
        if (r.start <= current.end + 1) {
          current.end = Math.max(current.end, r.end);
        } else {
          mergedRanges.push({ document_id: documentId, start_index: current.start, end_index: current.end });
          current = r;
        }
      }
      mergedRanges.push({ document_id: documentId, start_index: current.start, end_index: current.end });
    }

    const neighbors = await this.db.getChunksInRanges(mergedRanges);

    // Merge: keep max similarity for chunks that were originally retrieved.
    const merged = new Map<string, SearchResult>();
    for (const n of neighbors) merged.set(n.chunk_id, n);
    for (const c of chunks) {
      const existing = merged.get(c.chunk_id);
      merged.set(c.chunk_id, existing ? { ...existing, similarity: Math.max(existing.similarity, c.similarity) } : c);
    }

    // Sort: documents by best match similarity, then chunk_index asc.
    const docScore = new Map<string, number>();
    for (const c of chunks) {
      const prev = docScore.get(c.document_id) ?? 0;
      docScore.set(c.document_id, Math.max(prev, c.similarity));
    }

    const docs = [...new Set([...merged.values()].map(c => c.document_id))].sort(
      (a, b) => (docScore.get(b) ?? 0) - (docScore.get(a) ?? 0)
    );

    const byDocChunks = new Map<string, SearchResult[]>();
    for (const c of merged.values()) {
      const list = byDocChunks.get(c.document_id) || [];
      list.push(c);
      byDocChunks.set(c.document_id, list);
    }

    const result: SearchResult[] = [];
    for (const docId of docs) {
      const list = byDocChunks.get(docId) || [];
      list.sort((a, b) => a.chunk_index - b.chunk_index);
      result.push(...list);
    }

    return result;
  }
}

File: api/backend/services/neighbor_chunk_service.py
from __future__ import annotations

from backend.models.database import DatabaseModel, SearchResult


def add_neighbor_chunks(top_chunks: list[dict], all_chunks: list[dict], window: int = 1) -> list[dict]:
    chunk_map: dict[str, dict] = {}
    for chunk in all_chunks:
        key = f"{chunk.get('docId')}_{chunk.get('order')}"
        chunk_map[key] = chunk

    result: dict[str, dict] = {}
    for chunk in top_chunks:
        for offset in range(-window, window + 1):
            key = f"{chunk.get('docId')}_{int(chunk.get('order', 0)) + offset}"
            neighbor = chunk_map.get(key)
            if neighbor:
                result[str(neighbor.get("chunkId"))] = neighbor

    return list(result.values())


class NeighborChunkService:
    def __init__(self, db: DatabaseModel) -> None:
        self.db = db

    def expand(self, chunks: list[SearchResult], window: int = 1) -> list[SearchResult]:
        width = max(0, int(window))
        if width == 0 or not chunks:
            return chunks

        by_doc: dict[str, list[tuple[int, int]]] = {}
        for chunk in chunks:
            start = max(0, chunk.chunk_index - width)
            end = chunk.chunk_index + width
            by_doc.setdefault(chunk.document_id, []).append((start, end))

        merged_ranges: list[dict[str, int | str]] = []
        for document_id, ranges in by_doc.items():
            ranges.sort(key=lambda item: item[0])
            current_start, current_end = ranges[0]

            for next_start, next_end in ranges[1:]:
                if next_start <= current_end + 1:
                    current_end = max(current_end, next_end)
                else:
                    merged_ranges.append(
                        {
                            "document_id": document_id,
                            "start_index": current_start,
                            "end_index": current_end,
                        }
                    )
                    current_start, current_end = next_start, next_end

            merged_ranges.append(
                {
                    "document_id": document_id,
                    "start_index": current_start,
                    "end_index": current_end,
                }
            )

        try:
            neighbors = self.db.get_chunks_in_ranges(merged_ranges)
        except Exception as exc:
            message = str(exc).lower()
            if "index required" in message or "payload index" in message:
                return chunks
            raise

        merged: dict[str, SearchResult] = {item.chunk_id: item for item in neighbors}
        for chunk in chunks:
            existing = merged.get(chunk.chunk_id)
            if existing:
                existing.similarity = max(existing.similarity, chunk.similarity)
            else:
                merged[chunk.chunk_id] = chunk

        doc_score: dict[str, float] = {}
        for chunk in chunks:
            doc_score[chunk.document_id] = max(doc_score.get(chunk.document_id, 0.0), chunk.similarity)

        docs = sorted({item.document_id for item in merged.values()}, key=lambda doc: doc_score.get(doc, 0.0), reverse=True)

        by_doc_chunks: dict[str, list[SearchResult]] = {}
        for item in merged.values():
            by_doc_chunks.setdefault(item.document_id, []).append(item)

        output: list[SearchResult] = []
        for document_id in docs:
            values = by_doc_chunks.get(document_id, [])
            values.sort(key=lambda item: item.chunk_index)
            output.extend(values)

        return output
